addons/mail/static/src/discuss/discuss.js

```diff
@@ -31,6 +52,83 @@ export class Discuss extends Component {
     startCall() {
         this.messaging.startCall(this.messaging.discuss.threadId);
     }
+
+    /**
+     * @param {MouseEvent} ev
+     */
+    onClickThreadName(ev) {
+        if (!this.currentThread().isRenameable()) {
+            return;
+        }
+        this.state.doFocusOnThreadNameInput = true;
+        this.state.isEditingThreadName = true;
+        this.state.isMouseOverThreadName = false;
+        this.state.pendingThreadName = this.currentThread().name;
+    }
+
+    onInputThreadNameInput() {
+        this.state.pendingThreadName = this.threadNameInputRef.el.value.trim();
+    }
+
+    /**
+     * @param {KeyboardEvent} ev
+     */
+    onKeyDownThreadNameInput(ev) {
+        switch (ev.key) {
+            case 'Enter':
+                this._applyThreadRename();
+                break;
+            case 'Escape':
+                this._discardThreadRename();
+                break;
+        }
+    }
+
+    /**
+     * @param {MouseEvent} ev
+     */
+    onMouseEnterThreadName(ev) {
+        if (!this.currentThread().isRenameable()) {
+            return;
+        }
+        this.state.isMouseOverThreadName = true;
+    }
+
+    /**
+     * @param {MouseEvent} ev
+     */
+    onMouseLeaveThreadName(ev) {
+        this.state.isMouseOverThreadName = false;
+    }
+
+    async _applyThreadRename() {
+        const newName = this.state.pendingThreadName;
+        const thread = this.currentThread();
+        if (
+            thread.type === 'chat' && newName !== thread.name ||
+            thread.type === 'channel' && newName && newName !== thread.name ||
+            thread.type === 'group' && newName !== thread.name
+        ) {
+            thread.name = newName;
```

Comment:
in my opinion, this line should be moved into the `setThreadName` method in messaging